db_stress_tool/expected_state.cc

```diff
@@ -196,23 +252,325 @@ Status FileExpectedStateManager::Open() {
+    s = db->StartTrace(TraceOptions(), std::move(trace_writer));
+  }
+
+  // Delete old state/trace files. Deletion order does not matter since we only
+  // delete after successfully saving new files, so old files will never be used
+  // again, even if we crash.
+  if (s.ok()) {
+    if (old_saved_seqno != kMaxSequenceNumber &&
+        old_saved_seqno != saved_seqno_) {
+      s = Env::Default()->DeleteFile(
+          GetPathForFilename(ToString(old_saved_seqno) + kStateFilenameSuffix));
+    }
+  }
+  if (s.ok()) {
+    if (old_saved_seqno != kMaxSequenceNumber &&
+        old_saved_seqno != saved_seqno_) {
+      s = Env::Default()->DeleteFile(
+          GetPathForFilename(ToString(old_saved_seqno) + kTraceFilenameSuffix));
+    }
+  }
+  return s;
+}
+#else   // ROCKSDB_LITE
+Status FileExpectedStateManager::SaveAtAndAfter(DB* /* db */) {
+  return Status::NotSupported();
+}
+#endif  // ROCKSDB_LITE
+
+bool FileExpectedStateManager::HasHistory() {
+  return saved_seqno_ != kMaxSequenceNumber;
+}
+
+#ifndef ROCKSDB_LITE
+Status FileExpectedStateManager::Restore(DB* db) {
+  // An `ExpectedStateTraceRecordHandler` applies a configurable number of
+  // write operation trace records to the configured expected state.
+  class ExpectedStateTraceRecordHandler : public TraceRecord::Handler,
+                                          public WriteBatch::Handler {
+   public:
+    ExpectedStateTraceRecordHandler(uint64_t max_write_ops,
+                                    ExpectedState* state)
+        : max_write_ops_(max_write_ops), state_(state) {}
+
+    ~ExpectedStateTraceRecordHandler() {
+      assert(num_write_ops_ == max_write_ops_);
+    }
+
+    Status Handle(const WriteQueryTraceRecord& record,
+                  std::unique_ptr<TraceRecordResult>* /* result */) override {
+      if (num_write_ops_ == max_write_ops_) {
+        return Status::OK();
+      }
+      WriteBatch batch(record.GetWriteBatchRep().ToString());
+      return batch.Iterate(this);
+    }
+
+    // Ignore reads.
+    Status Handle(const GetQueryTraceRecord& /* record */,
+                  std::unique_ptr<TraceRecordResult>* /* result */) override {
+      return Status::OK();
+    }
+
+    // Ignore reads.
+    Status Handle(const IteratorSeekQueryTraceRecord& /* record */,
+                  std::unique_ptr<TraceRecordResult>* /* result */) override {
+      return Status::OK();
+    }
+
+    // Ignore reads.
+    Status Handle(const MultiGetQueryTraceRecord& /* record */,
+                  std::unique_ptr<TraceRecordResult>* /* result */) override {
+      return Status::OK();
+    }
+
+    // Below are the WriteBatch::Handler overrides. We could use a separate
+    // object, but it's convenient and works to share state with the
+    // `TraceRecord::Handler`.
+
+    Status PutCF(uint32_t column_family_id, const Slice& key,
+                 const Slice& value) override {
+      uint64_t key_id;
+      if (!GetIntVal(key.ToString(), &key_id)) {
```

Comment:
I'm going to try to ignore the inherited craziness of `GetIntVal` (extern inline taking a string by value, apparently decoding to unsigned where the encode encodes from signed, and builds a temporary vector rather than using values as they are decoded)